File: rust/tests/tests/test_meta_string.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use std::iter;

use fory_core::meta::{Encoding, MetaStringDecoder, MetaStringEncoder};

#[test]
fn test_encode_meta_string_lower_special() {
    let encoder = MetaStringEncoder::new();
    let bytes1 = encoder.encode_lower_special("abc_def").unwrap();
    assert_eq!(bytes1.len(), 5);
    let bytes2 = encoder
        .encode("org.apache.fory.benchmark.data")
        .unwrap()
        .bytes;
    assert_eq!(bytes2.len(), 19);
    let bytes3 = encoder.encode("MediaContent").unwrap().bytes;
    assert_eq!(bytes3.len(), 9);
    let decoder = MetaStringDecoder::new();
    assert_eq!(
        decoder.decode(&bytes1, Encoding::LowerSpecial).unwrap(),
        "abc_def"
    );
    for i in 0..128 {
        let origin_string: String = iter::repeat_with(|| {
            let char_a = b'a';
            ((char_a + i as u8) % 26 + char_a) as char
        })
        .take(i)
        .collect();
        let encoded = encoder.encode_lower_special(&origin_string).unwrap();
        let decoded = decoder.decode(&encoded, Encoding::LowerSpecial).unwrap();
        assert_eq!(decoded, origin_string);
    }
}

fn create_string(length: usize) -> String {
    (0..length)
        .map(|j| {
            let n = j % 64;
            match n {
                0..=25 => (b'a' + n as u8) as char,
                26..=51 => (b'A' + (n - 26) as u8) as char,
                52..=61 => (b'0' + (n - 52) as u8) as char,
                62 => '.',
                63 => '_',
                _ => unreachable!(),
            }
        })
        .collect()
}

#[test]
fn test_encode_meta_string_lower_upper_digit_special() {
    let encoder = MetaStringEncoder::new();
    let encoded = encoder
        .encode_lower_upper_digit_special("ExampleInput123")
        .unwrap();
    assert_eq!(encoded.len(), 12);

    let decoder = MetaStringDecoder::new();
    let decoded = decoder
        .decode(&encoded, Encoding::LowerUpperDigitSpecial)
        .unwrap();
    assert_eq!(decoded, "ExampleInput123");

    for i in 1..128 {
        let origin_string = create_string(i);
        let encoded = encoder
            .encode_lower_upper_digit_special(&origin_string)
            .unwrap();
        let decoded = decoder
            .decode(&encoded, Encoding::LowerUpperDigitSpecial)
            .unwrap();
        assert_eq!(decoded, origin_string);
    }
}

#[test]
fn test_meta_string() {
    let encoder = MetaStringEncoder::new();

    for i in 1..=127 {
        let origin_string = create_string(i);

        let meta_string = encoder.encode(&origin_string).unwrap();
        assert_ne!(meta_string.encoding, Encoding::Utf8);
        assert_eq!(meta_string.original, origin_string);

        let decoder = MetaStringDecoder::new();
        let new_string = decoder
            .decode(&meta_string.bytes, meta_string.encoding)
            .unwrap();
        assert_eq!(new_string, origin_string);
    }
}

#[test]
fn test_encode_empty_string() {
    let encoder = MetaStringEncoder::new();
    let decoder = MetaStringDecoder::new();
    for encoding in [
        Encoding::LowerSpecial,
        Encoding::LowerUpperDigitSpecial,
        Encoding::FirstToLowerSpecial,
        Encoding::AllToLowerSpecial,
        Encoding::Utf8,
    ] {
        let meta_string = encoder.encode_with_encoding("", encoding).unwrap();
        assert_eq!(meta_string.bytes.len(), 0);
        let decoded = decoder
            .decode(&meta_string.bytes, meta_string.encoding)
            .unwrap();
        assert_eq!(decoded, "");
    }
}

#[test]
fn test_encode_characters_outside_of_lower_special() {
    let encoder = MetaStringEncoder::new();
    let test_string = "abcdefABCDEF1234!@#";
    let meta_string = encoder.encode(test_string).unwrap();
    assert_eq!(meta_string.encoding, Encoding::Utf8);
}

#[test]
fn test_all_to_upper_special_encoding() {
    let encoder = MetaStringEncoder::new();
    let decoder = MetaStringDecoder::new();
    let test_string = "ABC_DEF";
    let meta_string = encoder.encode(test_string).unwrap();
    assert_eq!(meta_string.encoding, Encoding::LowerUpperDigitSpecial);
    let decoded_string = decoder
        .decode(&meta_string.bytes, meta_string.encoding)
        .unwrap();
    assert_eq!(decoded_string, test_string);
}

#[test]
fn test_first_to_lower_special_encoding() {
    let encoder = MetaStringEncoder::new();
    let decoder = MetaStringDecoder::new();
    let test_string = "Aabcdef";
    let meta_string = encoder.encode(test_string).unwrap();
    assert_eq!(meta_string.encoding, Encoding::FirstToLowerSpecial);
    let decoded_string = decoder
        .decode(&meta_string.bytes, meta_string.encoding)
        .unwrap();
    assert_eq!(decoded_string, test_string);
}

#[test]
fn test_utf8_encoding() {
    let encoder = MetaStringEncoder::new();
    let test_string = "你好，世界";
    let meta_string = encoder.encode(test_string).unwrap();
    assert_eq!(meta_string.encoding, Encoding::Utf8);
    let decoder = MetaStringDecoder::new();
    let decoded_string = decoder
        .decode(&meta_string.bytes, meta_string.encoding)
        .unwrap();
    assert_eq!(decoded_string, test_string);
    let test_string = "aA$";
    let meta_string = encoder.encode(test_string).unwrap();
    assert_eq!(meta_string.encoding, Encoding::Utf8);
    let decoder = MetaStringDecoder::new();
    let decoded_string = decoder
        .decode(&meta_string.bytes, meta_string.encoding)
        .unwrap();
    assert_eq!(decoded_string, test_string);
}

#[test]
fn test_strip_last_char() {
    let encoder = MetaStringEncoder::new();
    let test_string = "abc";
    let encoded_meta_string = encoder.encode(test_string).unwrap();
    assert!(!encoded_meta_string.strip_last_char);

    let test_string = "abcde";
    let encoded_meta_string = encoder.encode(test_string).unwrap();
    assert!(encoded_meta_string.strip_last_char);
}

#[test]
fn test_empty_string() {
    let encoder = MetaStringEncoder::new();
    let decoder = MetaStringDecoder::new();
    let meta_string = encoder.encode("").unwrap();
    assert!(meta_string.bytes.is_empty());
    let decoded = decoder
        .decode(&meta_string.bytes, meta_string.encoding)
        .unwrap();
    assert_eq!(decoded, "");
}

#[test]
fn test_ascii_encoding() {
    let encoder = MetaStringEncoder::new();
    let test_string = "asciiOnly";
    let encoded_meta_string = encoder.encode(test_string).unwrap();
    assert_ne!(encoded_meta_string.encoding, Encoding::Utf8);
    assert_eq!(encoded_meta_string.encoding, Encoding::AllToLowerSpecial);
}

#[test]
fn test_non_ascii_encoding() {
    let encoder = MetaStringEncoder::new();
    let test_string = "こんにちは";
    let encoded_meta_string = encoder.encode(test_string).unwrap();
    assert_eq!(encoded_meta_string.encoding, Encoding::Utf8);
}

#[test]
fn test_non_ascii_encoding_and_non_utf8() {
    let encoder = MetaStringEncoder::new();
    let non_ascii_string = "こんにちは";

    match encoder.encode_with_encoding(non_ascii_string, Encoding::LowerSpecial) {
        Err(err) => {
            assert_eq!(
                err.to_string(),
                "Non-ASCII characters in meta string are not allowed"
            );
        }
        Ok(_) => panic!("Expected an error due to non-ASCII character with non-UTF8 encoding"),
    }
}

File: python/pyfury/tests/test_metastring.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import pytest
from pyfury.meta.metastring import (
    MetaStringEncoder,
    MetaStringDecoder,
    Encoding,
)


def test_encode_metastring_lower_special():
    encoder = MetaStringEncoder(special_char1=".", special_char2="_")
    decoder = MetaStringDecoder(special_char1=".", special_char2="_")

    # Test for encoding and decoding
    encoded = encoder._encode_lower_special("abc_def")
    assert len(encoded) == 5
    assert len(encoder.encode("org.apache.fury.benchmark.data").encoded_data) == 19
    assert len(encoder.encode("MediaContent").encoded_data) == 9
    decoded = decoder.decode(encoded, Encoding.LOWER_SPECIAL)
    assert decoded == "abc_def"

    for i in range(128):
        builder = "".join(chr(ord("a") + j % 26) for j in range(i))
        encoded = encoder._encode_lower_special(builder)
        decoded = decoder.decode(encoded, Encoding.LOWER_SPECIAL)
        assert decoded == builder


def test_encode_metastring_lower_upper_digit_special():
    encoder = MetaStringEncoder(special_char1=".", special_char2="_")
    decoder = MetaStringDecoder(special_char1=".", special_char2="_")

    # Test for encoding and decoding
    encoded = encoder._encode_lower_upper_digit_special("ExampleInput123")
    assert len(encoded) == 12
    decoded = decoder.decode(encoded, Encoding.LOWER_UPPER_DIGIT_SPECIAL)
    assert decoded == "ExampleInput123"

    for i in range(1, 128):
        string = create_string(i)
        encoded = encoder._encode_lower_upper_digit_special(string)
        decoded = decoder.decode(encoded, Encoding.LOWER_UPPER_DIGIT_SPECIAL)
        assert decoded == string, f"Failed at {i}"


def create_string(length):
    chars = []
    for j in range(length):
        n = j % 64
        if n < 26:
            chars.append(chr(ord("a") + n))
        elif n < 52:
            chars.append(chr(ord("A") + (n - 26)))
        elif n < 62:
            chars.append(chr(ord("0") + (n - 52)))
        elif n == 62:
            chars.append(".")
        else:
            chars.append("_")
    return "".join(chars)


def test_metastring():
    encoder = MetaStringEncoder(special_char1=".", special_char2="_")
    decoder = MetaStringDecoder(special_char1=".", special_char2="_")

    for i in range(1, 128):
        try:
            string = create_string(i)
            metastring = encoder.encode(string)
            assert metastring.encoding != Encoding.UTF_8
            assert metastring.original == string

            new_string = decoder.decode(metastring.encoded_data, metastring.encoding)
            assert new_string == string
        except Exception as e:
            pytest.fail(f"Failed at {i} with exception: {str(e)}")


def test_encode_empty_string():
    encoder = MetaStringEncoder(special_char1=".", special_char2="_")
    decoder = MetaStringDecoder(special_char1=".", special_char2="_")

    for encoding in [
        Encoding.LOWER_SPECIAL,
        Encoding.LOWER_UPPER_DIGIT_SPECIAL,
        Encoding.FIRST_TO_LOWER_SPECIAL,
        Encoding.ALL_TO_LOWER_SPECIAL,
        Encoding.UTF_8,
    ]:
        metastring = encoder.encode_with_encoding("", encoding)
        assert len(metastring.encoded_data) == 0
        decoded = decoder.decode(metastring.encoded_data, metastring.encoding)
        assert decoded == ""


def test_encode_characters_outside_of_lower_special():
    encoder = MetaStringEncoder(special_char1=".", special_char2="_")

    test_string = "abcdefABCDEF1234!@#"
    metastring = encoder.encode(test_string)
    assert metastring.encoding == Encoding.UTF_8


def test_all_to_upper_special_encoding():
    encoder = MetaStringEncoder(special_char1=".", special_char2="_")
    decoder = MetaStringDecoder(special_char1=".", special_char2="_")

    test_string = "ABC_DEF"
    metastring = encoder.encode(test_string)
    assert metastring.encoding == Encoding.LOWER_UPPER_DIGIT_SPECIAL
    decoded_string = decoder.decode(metastring.encoded_data, metastring.encoding)
    assert decoded_string == test_string


def test_first_to_lower_special_encoding():
    encoder = MetaStringEncoder(special_char1=".", special_char2="_")
    decoder = MetaStringDecoder(special_char1=".", special_char2="_")

    test_string = "Aabcdef"
    metastring = encoder.encode(test_string)
    assert metastring.encoding == Encoding.FIRST_TO_LOWER_SPECIAL
    decoded_string = decoder.decode(metastring.encoded_data, metastring.encoding)
    assert decoded_string == test_string


def test_utf8_encoding():
    encoder = MetaStringEncoder(special_char1=".", special_char2="_")
    decoder = MetaStringDecoder(special_char1=".", special_char2="_")

    test_string = "你好，世界"  # Non-Latin characters
    metastring = encoder.encode(test_string)
    assert metastring.encoding == Encoding.UTF_8
    decoded_string = decoder.decode(metastring.encoded_data, metastring.encoding)
    assert decoded_string == test_string


def test_strip_last_char():
    encoder = MetaStringEncoder(special_char1=".", special_char2="_")

    test_string = "abc"  # encoded as 1|00000|00, 001|00010, exactly two bytes
    encoded_metastring = encoder.encode(test_string)
    assert not encoded_metastring.strip_last_char

    test_string = "abcde"  # encoded as 1|00000|00, 001|00010, 00011|001, 00xxxxxx, stripped last char
    encoded_metastring = encoder.encode(test_string)
    assert encoded_metastring.strip_last_char


def test_empty_string():
    encoder = MetaStringEncoder(special_char1=".", special_char2="_")
    decoder = MetaStringDecoder(special_char1=".", special_char2="_")

    metastring = encoder.encode("")
    assert metastring.encoded_data == bytes()

    decoded = decoder.decode(metastring.encoded_data, metastring.encoding)
    assert decoded == ""


def test_ascii_encoding():
    encoder = MetaStringEncoder(special_char1=".", special_char2="_")

    test_string = "asciiOnly"
    encoded_metastring = encoder.encode(test_string)
    assert encoded_metastring.encoding != Encoding.UTF_8
    assert encoded_metastring.encoding == Encoding.ALL_TO_LOWER_SPECIAL


def test_non_ascii_encoding():
    encoder = MetaStringEncoder(special_char1=".", special_char2="_")

    test_string = "こんにちは"  # Non-ASCII string
    encoded_metastring = encoder.encode(test_string)
    assert encoded_metastring.encoding == Encoding.UTF_8


def test_non_ascii_encoding_and_non_utf8():
    encoder = MetaStringEncoder(special_char1=".", special_char2="_")

    non_ascii_string = "こんにちは"  # Non-ASCII string

    with pytest.raises(
        ValueError, match="Unsupported character for LOWER_SPECIAL encoding: こ"
    ):
        encoder.encode_with_encoding(non_ascii_string, Encoding.LOWER_SPECIAL)
